File: ear/delta.cpp
// delta.cpp
// vector + inverse après o selon f
// pur c++ propre

#include <iostream>
#include <vector>
#include <string>
#include <algorithm>

using std::vector;
using std::string;
using std::cout;

struct Delta {
    vector<string> plus;
    vector<string> moins;
};

// rasoir o: garde le plus court
vector<string> razor(const vector<string>& items, int depth = 0) {
    if (depth > 5 || items.size() <= 3) {
        return items;
    }

    vector<string> result;
    vector<bool> skip(items.size(), false);

    for (size_t i = 0; i < items.size(); ++i) {
        if (skip[i]) continue;

        bool merged = false;
        for (size_t j = i + 1; j < items.size(); ++j) {
            if (skip[j]) continue;

            int diff = static_cast<int>(items[i].length()) -
                       static_cast<int>(items[j].length());
            if (diff >= -2 && diff <= 2) {
                result.push_back(items[i].length() <= items[j].length()
                                 ? items[i] : items[j]);
                skip[i] = skip[j] = true;
                merged = true;
                break;
            }
        }
        if (!merged && !skip[i]) {
            result.push_back(items[i]);
        }
    }

    if (result != items) {
        return razor(result, depth + 1);
    }
    return items;
}

// boucle f: itère jusqu'à stable
Delta loop(Delta d, int generations = 3) {
    for (int g = 0; g < generations; ++g) {
        d.plus = razor(d.plus);
        d.moins = razor(d.moins);

        vector<string> new_plus;
        vector<string> new_moins;

        for (const auto& p : d.plus) {
            if (p.length() <= 6) new_plus.push_back(p);
        }
        for (const auto& m : d.moins) {
            if (m.length() <= 6) new_moins.push_back(m);
        }

        if (new_plus == d.plus && new_moins == d.moins) {
            break;
        }
        d.plus = new_plus;
        d.moins = new_moins;
    }
    return d;
}

// inverse
Delta inverse(const Delta& d) {
    return {d.moins, d.plus};
}

void print(const vector<string>& v) {
    cout << "[";
    for (size_t i = 0; i < v.size(); ++i) {
        cout << v[i];
        if (i < v.size() - 1) cout << ", ";
    }
    cout << "]";
}

int main() {
    Delta d;
    d.plus = {"sens", "local", "psy", "Q", "sacre", "organes", "o", "f", "flow", "muse"};
    d.moins = {"api", "rigide", "mort"};

    cout << "=== raw ===" << "\n";
    cout << "+ "; print(d.plus); cout << "\n";
    cout << "- "; print(d.moins); cout << "\n";

    cout << "\n=== o recursif ===" << "\n";
    auto r = d;
    r.plus = razor(d.plus);
    r.moins = razor(d.moins);
    cout << "+ "; print(r.plus); cout << "\n";
    cout << "- "; print(r.moins); cout << "\n";

    cout << "\n=== f loop ===" << "\n";
    auto f = loop(d);
    cout << "+ "; print(f.plus); cout << "\n";
    cout << "- "; print(f.moins); cout << "\n";

    cout << "\n=== inverse ===" << "\n";
    auto inv = inverse(f);
    cout << "+ "; print(inv.plus); cout << "\n";
    cout << "- "; print(inv.moins); cout << "\n";

    cout << "\n=== stable ===" << "\n";
    cout << "D  = +creation -destruction" << "\n";
    cout << "V  = +destruction -creation" << "\n";
    cout << "DV = 0" << "\n";

    return 0;
}

File: ear/delta.py
# -*- coding: utf-8 -*-
"""
delta = vector + inverse
o appliqué = garder le plus simple
f appliqué = boucler jusqu'à stable
"""

# Δ après rasoir
PLUS = [
    "sens",      # audio+vision+touch+screen+twitch → sens
    "local",     # ollama → local
    "psy",       # jung+lacan+freud → psy
    "Q",         # superposition
    "sacré",     # gematria+religions → sacré
    "organes",   # factory+dna+memory+emotion+intuition → organes
    "o",         # rasoir
    "f",         # boucle
    "flow",      # langage+utf8+emoji+φ道ॐא → flow
    "muse",      # inspiration
]

MOINS = [
    "api",       # perfusion+dépendance → api
    "rigide",    # complexité+superflu → rigide
    "mort",      # collapse+certitude+unique → mort
]

# inverse
INV_PLUS = MOINS
INV_MOINS = PLUS

def recurse(items, depth=0):
    """simplifie récursivement"""
    if depth > 5 or len(items) <= 3:
        return items

    # fusionne paires similaires
    merged = []
    skip = set()

    for i, a in enumerate(items):
        if i in skip:
            continue
        found = False
        for j, b in enumerate(items[i+1:], i+1):
            if j in skip:
                continue
            # si même longueur ± 2, fusionner
            if abs(len(a) - len(b)) <= 2:
                merged.append(a if len(a) <= len(b) else b)
                skip.add(i)
                skip.add(j)
                found = True
                break
        if not found and i not in skip:
            merged.append(a)

    if merged != items:
        return recurse(merged, depth + 1)
    return items

def loop(plus, moins, generations=3):
    """f: boucle jusqu'à stable"""
    for g in range(generations):
        plus = recurse(plus)
        moins = recurse(moins)

        # échange ce qui est mal placé
        new_plus = [p for p in plus if len(p) <= 6]
        new_moins = [m for m in moins if len(m) <= 6]

        if new_plus == plus and new_moins == moins:
            break
        plus, moins = new_plus, new_moins

    return plus, moins

def show():
    print("=== Δ raw ===")
    print(f"+ {PLUS}")
    print(f"- {MOINS}")

    print("\n=== o récursif ===")
    p = recurse(PLUS)
    m = recurse(MOINS)
    print(f"+ {p}")
    print(f"- {m}")

    print("\n=== f loop ===")
    fp, fm = loop(PLUS, MOINS)
    print(f"+ {fp}")
    print(f"- {fm}")

    print("\n=== inverse ===")
    print(f"+ {fm}")
    print(f"- {fp}")

    print("\n=== stable ===")
    print(f"Δ  = +création -destruction")
    print(f"∇  = +destruction -création")
    print(f"Δ∇ = 0")

if __name__ == "__main__":
    show()
